extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/DefaultTokenIntrospectionUserInfoCache.java

```diff
@@ -0,0 +1,186 @@
+                    // Remove all the entries which have expired
+                    removeInvalidEntries(true);
+                }
+            };
+            cleanUpTimer = new Timer("OIDC introspection cache cleanup timer");
+            cleanUpTimer.scheduleAtFixedRate(cleanUpTask, 3000, cleanUpTimerInterval.get().toMillis());
+        }
+    }
+
+    @PreDestroy
+    public void stop() {
+        if (cleanUpTimer != null) {
+            cleanUpTimer.cancel();
+            cleanUpTimer = null;
+        }
+    }
+
+    @Override
+    public Uni<Void> addIntrospection(String token, TokenIntrospection introspection, OidcTenantConfig oidcConfig,
+            AddIntrospectionRequestContext requestContext) {
+        if (maxSize > 0) {
+            CacheEntry entry = findValidCacheEntry(token);
+            if (entry != null) {
+                entry.introspection = introspection;
+            } else if (prepareSpaceForNewCacheEntry()) {
+                cacheMap.put(token, new CacheEntry(introspection));
+            }
+        }
+
+        return CodeAuthenticationMechanism.VOID_UNI;
+    }
+
+    @Override
+    public Uni<TokenIntrospection> getIntrospection(String token, OidcTenantConfig oidcConfig,
+            GetIntrospectionRequestContext requestContext) {
+        CacheEntry entry = findValidCacheEntry(token);
+        return entry == null ? NULL_INTROSPECTION_UNI : Uni.createFrom().item(entry.introspection);
+    }
+
+    @Override
+    public Uni<Void> addUserInfo(String token, UserInfo userInfo, OidcTenantConfig oidcConfig,
+            AddUserInfoRequestContext requestContext) {
+        if (maxSize > 0) {
+            CacheEntry entry = findValidCacheEntry(token);
+            if (entry != null) {
+                entry.userInfo = userInfo;
+            } else if (prepareSpaceForNewCacheEntry()) {
+                cacheMap.put(token, new CacheEntry(userInfo));
+            }
+        }
+
+        return CodeAuthenticationMechanism.VOID_UNI;
+    }
+
+    @Override
+    public Uni<UserInfo> getUserInfo(String token, OidcTenantConfig oidcConfig,
+            GetUserInfoRequestContext requestContext) {
+        CacheEntry entry = findValidCacheEntry(token);
+        return entry == null ? NULL_USERINFO_UNI : Uni.createFrom().item(entry.userInfo);
+    }
+
+    public int getCacheSize() {
+        return cacheMap.size();
+    }
+
+    public void clearCache() {
+        cacheMap.clear();
+    }
+
+    private void removeInvalidEntries(boolean removeAll) {
+        long now = now();
+        for (Iterator<Map.Entry<String, CacheEntry>> it = cacheMap.entrySet().iterator(); it.hasNext();) {
+            Map.Entry<String, CacheEntry> next = it.next();
+            if (isEntryExpired(next.getValue(), now)) {
+                it.remove();
+                if (!removeAll) {
+                    break;
+                }
+            }
+        }
+    }
+
+    private boolean prepareSpaceForNewCacheEntry() {
```

Comment:
@stuartwdouglas It is racy, but I was not intending to have a strict enforcement of max-size, example, if it is 1000 then even if it will be full again after the check and it will overflow a bit, then we'd have say 1003, but from then on, until it drops down back to around 1000, no new entries will be added, correct ? Though I can give it a go with `AtomicInteger`